File: app/src/main/java/de/westnordost/osmagent/quests/OsmQuest.java
package de.westnordost.osmagent.quests;

import javax.annotation.Nullable;

import de.westnordost.osmagent.quests.types.QuestType;
import de.westnordost.osmapi.map.data.Element;

/** Represents one task for the user to complete/correct the data based on one OSM element */
public class OsmQuest implements Quest
{
	public OsmQuest(long id, QuestType type, long elementId, Element.Type elementType,
					QuestStatus status,	@Nullable StringMapChanges changes, ElementGeometry geometry)
	{
		this.id = id;
		this.type = type;
		this.elementId = elementId;
		this.elementType = elementType;
		this.changes = changes;
		this.geometry = geometry;
		this.status = status;
	}

	private long id;
	private final QuestType type;
	private QuestStatus status;
	private final ElementGeometry geometry;

	// underlying OSM data
	private final long elementId;
	private final Element.Type elementType;

	// and the changes to the tags (in the future, osmagent should probably be able to edit more
	// than just tags -> osmchange?)
	private StringMapChanges changes;

	public long getId()
	{
		return id;
	}

	public long getElementId()
	{
		return elementId;
	}

	public Element.Type getElementType()
	{
		return elementType;
	}

	@Override public QuestType getType()
	{
		return type;
	}

	public ElementGeometry getGeometry()
	{
		return geometry;
	}

	public StringMapChanges getChanges()
	{
		return changes;
	}

	public void setChanges(StringMapChanges changes)
	{
		this.changes = changes;
	}

	@Override public QuestStatus getStatus()
	{
		return status;
	}

	@Override public void setStatus(QuestStatus status)
	{
		this.status = status;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquest/OsmQuest.kt
package de.westnordost.streetcomplete.data.osm.osmquest

import java.util.Date

import de.westnordost.streetcomplete.data.quest.Quest
import de.westnordost.streetcomplete.data.quest.QuestStatus
import de.westnordost.streetcomplete.data.osm.changes.StringMapChanges
import de.westnordost.streetcomplete.data.quest.QuestType
import de.westnordost.osmapi.map.data.Element
import de.westnordost.osmapi.map.data.LatLon
import de.westnordost.streetcomplete.data.osm.elementgeometry.ElementGeometry
import de.westnordost.streetcomplete.data.osm.elementgeometry.ElementPolylinesGeometry
import de.westnordost.streetcomplete.data.osm.upload.HasElementTagChanges
import de.westnordost.streetcomplete.data.osm.upload.UploadableInChangeset
import de.westnordost.streetcomplete.util.measuredLength
import de.westnordost.streetcomplete.util.pointOnPolylineFromEnd
import de.westnordost.streetcomplete.util.pointOnPolylineFromStart

/** Represents one task for the user to complete/correct the data based on one OSM element  */
data class OsmQuest(
    override var id: Long?,
    override val osmElementQuestType: OsmElementQuestType<*>, // underlying OSM data
    override val elementType: Element.Type,
    override val elementId: Long,
    override var status: QuestStatus,
    override var changes: StringMapChanges?,
    var changesSource: String?,
    override var lastUpdate: Date,
    override val geometry: ElementGeometry
) : Quest, UploadableInChangeset, HasElementTagChanges {

    constructor(type: OsmElementQuestType<*>, elementType: Element.Type, elementId: Long, geometry: ElementGeometry)
        : this(null, type, elementType, elementId, QuestStatus.NEW, null, null, Date(), geometry)

    override val center: LatLon get() = geometry.center
    override val type: QuestType<*> get() = osmElementQuestType
    override val position: LatLon get() = center

    override val markerLocations: Collection<LatLon> get() {
        if (osmElementQuestType.hasMarkersAtEnds && geometry is ElementPolylinesGeometry) {
            val polyline = geometry.polylines[0]
            val length = polyline.measuredLength()
            if (length > 15 * 4) {
                return listOf(
                    polyline.pointOnPolylineFromStart(15.0)!!,
                    polyline.pointOnPolylineFromEnd(15.0)!!
                )
            }
        }
        return listOf(center)
    }

    override fun isApplicableTo(element: Element) = osmElementQuestType.isApplicableTo(element)

    /* --------------------------- UploadableInChangeset --------------------------- */

    override val source: String get() = changesSource!!
}
